### PR TITLE
BUG: Fix multiclass onehot inverse

### LINKED ISSUE #188: Issue with model setup
I am having issues getting SciKeras getting to work as expected; 

From evaluating the result it seems that SciKeras learns as expected (at least shown in verbose output) but does not modify the model (as predictions made by this model are all wrong).

I used this same model directly with Keras and got a reasonable score.

I am not sure if this is an installation issue (running this on Conda through pip v0.2.1), or if I overlooked a parameter while setting it up for my use case.
I know I left this a bit dry, I would appreciate some help knowing what to look for, and let me know what I can provide.

### Imports

```python
# import rfpimp
import pickle
import dask.dataframe as dd

# Custom Implementation
from modules.classification import Classification, ConfusionMatrix, RFVisualisation
```

```python
from tensorflow import keras
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import train_test_split, cross_val_score
import numpy as np
```

### Running Quickstart Example seems to work as expected...

```python
import numpy as np
from sklearn.datasets import make_classification
from tensorflow import keras

from scikeras.wrappers import KerasClassifier


X, y = make_classification(1000, 20, n_informative=10, random_state=0)
X = X.astype(np.float32)
y = y.astype(np.int64)

def get_model(hidden_layer_dim, meta):
    # note that meta is a special argument that will be
    # handed a dict containing input metadata
    n_features_in_ = meta["n_features_in_"]
    X_shape_ = meta["X_shape_"]
    n_classes_ = meta["n_classes_"]

    model = keras.models.Sequential()
    model.add(keras.layers.Dense(n_features_in_, input_shape=X_shape_[1:]))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(hidden_layer_dim))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(n_classes_))
    model.add(keras.layers.Activation("softmax"))
    return model

clf = KerasClassifier(
    get_model,
    loss="sparse_categorical_crossentropy",
    hidden_layer_dim=100,
)

clf.fit(X, y)
y_proba = clf.predict_proba(X)
y_proba
```

    32/32 [==============================] - 1s 1ms/step - loss: 0.6739
    32/32 [==============================] - 0s 756us/step

    array([[0.4033311 , 0.59666884],
           [0.78341705, 0.21658295],
           [0.5785239 , 0.4214761 ],
           ...,
           [0.8525218 , 0.14747825],
           [0.5348476 , 0.46515232],
           [0.63776696, 0.36223304]], dtype=float32)

### Will run with my existing implementation...

```python
cl = Classification(verbose=True)

def evaluate_sample(featureset, description, classifier=default_classifier, 
                    sample=path['samples'], scaler=StandardScaler, 
                    validation=0.2, imbalance=None,
                    allcolumns=allcolumns, label='code', tileset='CVieL', 
                    nomenclature=nomenclature, resultspath=path['results'], 
                    verbose=True, overwrite=False):
    
    cl.vprint(f'Description:\t{description}')
    
    pkl_path = Path(resultspath, 'pickles', f'{description}.pickle')
    if not overwrite and pkl_path.is_file():
        cl.vprint('Run Exists! Loading from pickle...')
        return load_result_pickle(pkl_path)        
    
    cols = ['code']+[c for l in featureset for c in allcolumns if c.endswith(l)]
    featuresdf = cl.load_features(sample, cols)
    
    if imbalance is not None:
        featuresdf = cl.resample_features(featuresdf, imbalance)
    else:
        szdf = featuresdf.groupby(by='code').size()
        imbalance = round(szdf.max()/szdf.min(), 1)
            
    splits, featcols, validation = cl.split_features(featuresdf, label, validation, scaler)
    X_train, X_test, y_train, y_test = splits
    
    return splits

X_train, X_test, y_train, y_test = evaluate_sample(**{'featureset': o_ind_stats, 'description': 'S2 indices mean and variance with SciKeras'})
```

    Description:	S2 indices mean and variance with SciKeras
    Split Ratio:	0.2
    Split Shapes:	[(79997, 96), (20000, 96), (79997, 12), (20000, 12)]

### Using SciKeras, with failing scores

```python
# TODO: Fix SciKeras for scikit-learn integration
def fcn_model(hidden_layer_dim, meta):
    
    print('meta: {\n', '\n  '.join([f'{kv}' for kv in meta.items()]), '\n}')
    # note that meta is a special argument that will be
    # handed a dict containing input metadata
    n_features_in_ = meta["n_features_in_"]
    X_shape_ = meta["X_shape_"]
    n_classes_ = meta["n_classes_"]
    
    model = keras.models.Sequential()
    model.add(keras.layers.Dense(n_features_in_, input_shape=X_shape_[1:]))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(hidden_layer_dim))
    model.add(keras.layers.Activation("relu"))
    model.add(keras.layers.Dense(n_classes_))
    model.add(keras.layers.Activation("softmax"))
    return model

def get_model_fcn(modelargs={}):
    return KerasClassifier(fcn_model, 
                           hidden_layer_dim=128, 
                           epochs=1,
                           optimizer='adam',
                           loss='categorical_crossentropy',
                           metrics=['accuracy'],
                           **modelargs)

model = get_model_fcn()
model.fit(X_train, y_train)

columns, report, true_matrix, pred_matrix = cl.classification_metrics(model, [X_train, X_test, y_train, y_test], [0, 0])
```

    meta: {
     ('classes_', array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]))
      ('target_type_', 'multilabel-indicator')
      ('y_dtype_', dtype('uint8'))
      ('y_ndim_', 2)
      ('X_dtype_', dtype('float32'))
      ('X_shape_', (79997, 96))
      ('n_features_in_', 96)
      ('target_encoder_', ClassifierLabelEncoder(loss='categorical_crossentropy'))
      ('n_classes_', 12)
      ('n_outputs_', 1)
      ('n_outputs_expected_', 1)
      ('feature_encoder_', FunctionTransformer()) 
    }
    2500/2500 [==============================] - 3s 1ms/step - loss: 1.2360 - accuracy: 0.5530
    625/625 [==============================] - 0s 686us/step
    Overall: 0.130  Kappa: 0.000
    Weighted F1-Score: 0.030

    /home/ad/abelat/tharen/.conda/envs/map/lib/python3.6/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
      _warn_prf(average, modifier, msg_start, len(result))
    /home/ad/abelat/tharen/.conda/envs/map/lib/python3.6/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
      _warn_prf(average, modifier, msg_start, len(result))
    /home/ad/abelat/tharen/.conda/envs/map/lib/python3.6/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
      _warn_prf(average, modifier, msg_start, len(result))

### Using same SciKeras model, directly with Keras, is working

```python
# Co-Opted from SciKeras Model Compilation
meta = {
    'classes_': np.array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]), 
    'target_type_': 'multilabel-indicator', 
    'y_dtype_': np.dtype('uint8'), 
    'y_ndim_': 2, 
    'X_dtype_': np.dtype('float32'), 
    'X_shape_': (79997, 96), 
    'n_features_in_': 96, 
    #'target_encoder_': ClassifierLabelEncoder(loss='categorical_crossentropy'), 
    'n_classes_': 12, 
    'n_outputs_': 1, 
    'n_outputs_expected_': 1, 
    #'feature_encoder_': FunctionTransformer()
}

model = fcn_model(128, meta)
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=1)
columns, report, true_matrix, pred_matrix = cl.classification_metrics(model, [X_train, X_test, y_train, y_test], [0, 0])
```

    meta: {
     ('classes_', array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]))
      ('target_type_', 'multilabel-indicator')
      ('y_dtype_', dtype('uint8'))
      ('y_ndim_', 2)
      ('X_dtype_', dtype('float32'))
      ('X_shape_', (79997, 96))
      ('n_features_in_', 96)
      ('n_classes_', 12)
      ('n_outputs_', 1)
      ('n_outputs_expected_', 1) 
    }
    2500/2500 [==============================] - 3s 1ms/step - loss: 1.2370 - accuracy: 0.5477
    Overall: 0.660  Kappa: 0.615
    Weighted F1-Score: 0.651

### Classification Metrics Implementation

```python
# Model Analysis Metrics
def classification_metrics(self, model, splits, score):
    X_train, X_test, y_train, y_test = splits
    columns, labels, predictions = y_train.columns, y_test.values, model.predict(X_test)

    # Generate Report using Test Set predictions
    labels_list = labels.argmax(axis=1)
    predictions_list = predictions.argmax(axis=1)

    report = classification_report(labels_list, predictions_list, 
                                   target_names=columns, output_dict=True)

    # Add Train Set imbalance to the report
    imbalance = y_train.sum(axis=0)
    imbalance = imbalance/min(imbalance)

    for code, code_imbalance in imbalance.iteritems():
        report[code]['imbalance'] = code_imbalance

    # Manual Values not provided by the classification report
    report['kappa'] = cohen_kappa_score(labels_list, predictions_list)
    report['cv'] = {}
    report['cv']['mean'] = np.mean(score)
    report['cv']['std'] = np.std(score)
    report['samples'] = {}
    report['samples']['train'] = len(splits[0])
    report['samples']['test'] = len(splits[-1])
    report['samples']['features'] = len(splits[0].columns)

    self.vprint('Overall: %.3f  Kappa: %.3f' % (report['accuracy'], report['kappa']))
    self.vprint('Weighted F1-Score: %.3f' % (report['weighted avg']['f1-score']))

    true_matrix = confusion_matrix(labels_list, predictions_list, normalize='true')
    pred_matrix = confusion_matrix(labels_list, predictions_list, normalize='pred')

    return columns, report, true_matrix, pred_matrix
```

### PR DESCRIPTION
**Pull request by adriangb**
Resolves #188 

One-hot encoded targets were not being processed correctly. The tests were checking for shapes and dtypes (which were correct), but the values were not.

This implements 2 things:
- 2 line fix for the bug in `transformers.py`
- Extensive test refactoring to test for estimator scores (to prevent this bug in the future).

### CODE REVIEW

File: scikeras/wrappers.py [hunk 1]
```diff
@@ -1626,6 +1626,9 @@ def r_squared(y_true, y_pred):
         """A simple Keras implementation of R^2 that can be used as a Keras
         loss function.
 
+        Note that this returns 1-R^2 so that it can be minimized as a loss
+        function.
```
Comment (stsievert) [inline scikeras/wrappers.py:1630]: I'm not sure I'd include this line. I think saying that R^2 is coefficient of determination suffices. Maybe you can say that R^2 is an equivalent measure to MSE (specifically with `mse = (1 - coef) * c` for some constant `c`).

File: scikeras/wrappers.py [hunk 2]
```diff
@@ -1641,6 +1644,6 @@ def r_squared(y_true, y_pred):
             tf.math.squared_difference(y_true, tf.math.reduce_mean(y_true, axis=0)),
             axis=0,
         )
-        return tf.math.reduce_mean(
+        return 1 - tf.math.reduce_mean(
```
Comment (stsievert) [inline scikeras/wrappers.py:1647]: Shouldn't tests break with this change? I don't think this function should be called "`r_squared`" if this value is returned.

File: scikeras/utils/transformers.py [hunk 3]
```diff
@@ -255,13 +254,12 @@ def inverse_transform(
                 class_predictions[:, idx] = 1
             class_predictions = self._final_encoder.inverse_transform(class_predictions)
         elif self._target_type == "multiclass-onehot":
-            # array([.8, .1, .1], [.1, .8, .1]) ->
-            # array([[1, 0, 0], [0, 1, 0]])
+            # array([.8, .1, .1], [.1, .8, .1]) -> array([[1, 0, 0], [0, 1, 0]])
             idx = np.argmax(y, axis=-1)
             class_predictions = np.zeros(y.shape, dtype=int)
-            class_predictions[:, idx] = 1
+            class_predictions[np.arange(idx.size), idx] = 1
```
Comment (stsievert) [inline scikeras/utils/transformers.py:260]: Style nit: why is this change necessary?

Comment (adriangb) [reply scikeras/utils/transformers.py:260]: This is the actual bug being fixed. Apparently `class_predictions[:, idx]` assigns 1 to every element in the array.

File: scikeras/wrappers.py (hunk 2 above)
Comment (adriangb) [reply scikeras/wrappers.py:1647]: There was a single test that "broke", it's "fixed" in `test_scoring.py`.

This said, and in relation to https://github.com/adriangb/scikeras/pull/189#discussion_r576316582, I've debated if this function should even be included in SciKeras. Do you think it's valuable to be able to use this as a loss function, as opposed to just using MSE? I don't think there is any real advantage, and thus this function should probably be removed. I forget why I added it in the first place; I'm guessing it was out of naivety.

File: scikeras/utils/transformers.py (hunk 3 above)
Comment (stsievert) [reply scikeras/utils/transformers.py:260]: Got it. That's because `x[:, [0, 1]] = 1` assigns the columns 0 and 1 to be 1 for every row. `x[idx0, idx1]` will assign `x[idx0[i], idx1[i]] = 1` for every `i` if `idx0` and `idx1` are the same length.

Style nit: maybe use `class_predictions[np.arange(len(idx)), idx] = 1` instead?

Comment (adriangb) [reply scikeras/utils/transformers.py:260]: Yeah that should work fine, I'll implement that change.

File: scikeras/wrappers.py (hunk 1 above)
Comment (adriangb) [reply scikeras/wrappers.py]: I resolved the comment from the docstring as per https://github.com/adriangb/scikeras/pull/189#discussion_r576321029. Can I resolve the comment?

File: scikeras/wrappers.py (hunk 2 above)
Comment (adriangb) [reply scikeras/wrappers.py]: I opened #197 to address this. All R^2 changes are reverted from this PR.

File: scikeras/utils/transformers.py (hunk 3 above)
Comment (adriangb) [reply scikeras/utils/transformers.py]: Changed as suggested.

**Comment by adriangb** [comment]
https://github.com/adriangb/scikeras/issues/188#issuecomment-779496955 confirms this fixes the original bug report. Will move forward with merge.
